graphjet-core/src/main/java/com/twitter/graphjet/algorithms/filters/TweetAuthorFilterBase.java

```diff
@@ -17,32 +17,43 @@
 package com.twitter.graphjet.algorithms.filters;
 
 import com.twitter.graphjet.algorithms.RecommendationRequest;
-import com.twitter.graphjet.bipartite.api.EdgeIterator;
 import com.twitter.graphjet.bipartite.LeftIndexedMultiSegmentBipartiteGraph;
-import com.twitter.graphjet.hashing.SmallArrayBasedLongToDoubleMap;
+import com.twitter.graphjet.bipartite.api.EdgeIterator;
 import com.twitter.graphjet.stats.StatsReceiver;
+
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
 
-public class TweetAuthorFilter extends ResultFilter {
-  private LongSet authoredTweets = new LongOpenHashSet();
+public abstract class TweetAuthorFilterBase extends ResultFilter {
+
+  private LongSet authoredTweets;
   private boolean isTweetAuthorsEmpty = false;
 
-  /**
-   * @param tweetAuthors the list of authors whose tweets will not be filtered. If left empty, no tweet will be filtered.
-   */
-  public TweetAuthorFilter(
+  public TweetAuthorFilterBase(
       LeftIndexedMultiSegmentBipartiteGraph leftIndexedBipartiteGraph,
       LongSet tweetAuthors,
-      StatsReceiver statsReceiver) {
+      StatsReceiver statsReceiver
+  ) {
     super(statsReceiver);
-    isTweetAuthorsEmpty = tweetAuthors.isEmpty();
-    generateAuthoredByUsersNodes(leftIndexedBipartiteGraph, tweetAuthors);
+    this.isTweetAuthorsEmpty = tweetAuthors.isEmpty();
+    this.authoredTweets = generateAuthoredByUsersNodes(leftIndexedBipartiteGraph, tweetAuthors);
+  }
+
+  public boolean isTweetAuthorsEmpty() {
+    return this.isTweetAuthorsEmpty;
   }
 
-  private void generateAuthoredByUsersNodes(
+  public boolean isTweetedByTweetAuthors(long tweetId) {
+    return authoredTweets.contains(tweetId);
+  }
+
+  /**
+   * Return the list of tweets authored by the input list of users
+   */
+  private LongSet generateAuthoredByUsersNodes(
       LeftIndexedMultiSegmentBipartiteGraph leftIndexedBipartiteGraph,
       LongSet tweetAuthors) {
+    LongSet authoredTweets = new LongOpenHashSet();
     for (long leftNode: tweetAuthors) {
       EdgeIterator edgeIterator = leftIndexedBipartiteGraph.getLeftNodeEdges(leftNode);
```

Comment:
This getLeftNodeEdges in the for loop is an expensive step. What it does is to retrieve all the engagements for a user, and since it happens inside a for loop, it will retrieve all the engagements for all users specified in tweetAuthors.

One alternative is to combine both whitelist and blacklist in the same filter so that we only iterate through all the edges once.

Reply:
The performance concern is totally fair, I had similar thoughts on this. However, combing the 2 filters feels a bit hacky, as the author sets for whitelist and blacklist could be different. I think it is the caller's responsibility to acknowledge the cost when choosing to use them. 

That said, the cheaper alternative to achieving this goal (filtering by author) is a social proof type filter, where we filter candidates out "tweet" social proofs. This way we avoid looping all author engagements. The drawback is, this filter applies to the entire seed user set, instead of a selection of users passed in by the caller like the blacklist author filter. The benefit is, it is a lot faster.